### PR TITLE
Update daemon JVM documentation

### PR DESCRIPTION
**Pull request by ljacomet**
Rendered version:

Screenshot 2024-05-31 at 18 08 52

Can be compared with https://docs.gradle.org/8.8-rc-2/userguide/gradle_daemon.html#sec:daemon_jvm_criteria

### CODE REVIEW

**Comment by big-guy** [comment]
@bot-gradle test bd

File: platforms/documentation/docs/src/docs/userguide/optimizing-performance/gradle_daemon.adoc [hunk 1]
```diff
@@ -171,43 +171,60 @@ You can also kill Daemons manually with your operating system.
...
-Using the _Daemon JVM discovery_ and _Daemon JVM criteria_ Gradle records the build's JVM requirements.
-This controls the JVM used to run Gradle itself.
+Building on the <<toolchains.adoc#toolchains,toolchain feature>>, you can now use declarative criteria to specify the JVM requirements for the build.
 
 === Daemon JVM criteria
 
-The _Daemon JVM criteria_ feature is similar to the `wrapper` task, which captures the current Gradle version.
-It uses the built-in `updateDaemonJvm` task to record the JVM requirements for the build based on the current JVM installation.
+The _daemon JVM criteria_ feature is built based on a task, similar to the <<gradle_wrapper.adoc#customizing_wrapper,`wrapper` one>>.
+The task, when run, will use its configuration, found in build script or as command line arguments, and will capture a JVM definition in the `gradle/daemon-jvm.properties` file.
```
Comment (big-guy) [inline platforms/documentation/docs/src/docs/userguide/optimizing-performance/gradle_daemon.adoc:188]:
```suggestion
The task uses its configuration in the build script or command-line arguments to capture a JVM definition in the `gradle/daemon-jvm.properties` file.
```

**Review by big-guy — APPROVED** [review]

**Review by lkasso — APPROVED** [review]

File: platforms/documentation/docs/src/docs/userguide/optimizing-performance/gradle_daemon.adoc [hunk 2]
```diff
@@ -171,43 +171,60 @@ You can also kill Daemons manually with your operating system.
...
-Using the _Daemon JVM discovery_ and _Daemon JVM criteria_ Gradle records the build's JVM requirements.
-This controls the JVM used to run Gradle itself.
+Building on the <<toolchains.adoc#toolchains,toolchain feature>>, you can now use declarative criteria to specify the JVM requirements for the build.
 
 === Daemon JVM criteria
 
-The _Daemon JVM criteria_ feature is similar to the `wrapper` task, which captures the current Gradle version.
-It uses the built-in `updateDaemonJvm` task to record the JVM requirements for the build based on the current JVM installation.
+The _daemon JVM criteria_ feature is built based on a task, similar to the <<gradle_wrapper.adoc#customizing_wrapper,`wrapper` one>>.
+The task uses its configuration in the build script or command-line arguments to capture a JVM definition in the `gradle/daemon-jvm.properties` file.
```
Comment (alllex) [inline platforms/documentation/docs/src/docs/userguide/optimizing-performance/gradle_daemon.adoc:188]:
```suggestion
The _daemon JVM criteria_ is controlled by a task, similarly to how <<gradle_wrapper.adoc#customizing_wrapper,`wrapper` task>> updates the wrapper properties.
When the task runs, it creates or updates the criteria in the `gradle/daemon-jvm.properties` file.
For more control, the task can be further configured in the build script or via command-line arguments.
```

File: platforms/documentation/docs/src/docs/userguide/optimizing-performance/gradle_daemon.adoc [hunk 3]
```diff
@@ -171,43 +171,60 @@ You can also kill Daemons manually with your operating system.
...
+include::sample[dir="snippets/daemon-jvm/customized-task/groovy",files="build.gradle[tags=customized-daemon-jvm-task]"]
+====
 
-On subsequent builds, _Daemon JVM discovery_ uses the recorded JVM requirements to locate a compatible JVM installation using the same underlying mechanisms as <<toolchains.adoc#toolchains, Java Toolchains>>.
-This ensures every person who starts the build will use the same JVM version.
+====
+.gradle/daemon-jvm.properties
+[source,properties]
+----
+#This file is generated by updateDaemonJvm
```
Comment (alllex) [inline platforms/documentation/docs/src/docs/userguide/optimizing-performance/gradle_daemon.adoc:203]:
```suggestion
# This file is generated by updateDaemonJvm
```

File: platforms/documentation/docs/src/docs/userguide/optimizing-performance/gradle_daemon.adoc [hunk 4]
```diff
@@ -171,43 +171,60 @@ You can also kill Daemons manually with your operating system.
...
+toolchainVersion=17
+----
+====
+
+Without editing the build script, you could produce the same output using the following command:
```
Comment (alllex) [inline platforms/documentation/docs/src/docs/userguide/optimizing-performance/gradle_daemon.adoc:208]:
```suggestion
The same properties file can be produced without configuring the task in the build script, and using a command-line argument instead:
```

File: platforms/documentation/docs/src/docs/userguide/optimizing-performance/gradle_daemon.adoc [hunk 5]
```diff
@@ -171,43 +171,60 @@ You can also kill Daemons manually with your operating system.
...
+
+----
+$ ./gradlew updateDaemonJvm --jvm-version=17
+----
 
-_Daemon JVM discovery_ does not support auto-provisioning of new JVM installations.
+If you run the task without any argument, then the JVM used by the daemon will provide the version value.
```
Comment (alllex) [inline platforms/documentation/docs/src/docs/userguide/optimizing-performance/gradle_daemon.adoc:214]:
```suggestion
If you run the task without any arguments and the properties file does not exist, then the JVM used by the daemon will provide the version value.
```

**Review by alllex — APPROVED** [review]
This is already much better, but here are a couple of suggestions to polish it further

File: platforms/documentation/docs/src/docs/userguide/optimizing-performance/gradle_daemon.adoc (hunk 3 above)
Comment (ljacomet) [reply platforms/documentation/docs/src/docs/userguide/optimizing-performance/gradle_daemon.adoc:203]: Yes, we should fix that, but that's a code change. I copied it from a real generated task

**Comment by ljacomet** [comment]
@bot-gradle merge

**Pull request merged**